be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsWithMapHandler.scala

```diff
@@ -0,0 +1,111 @@
+import ch.epfl.pop.model.objects.{Channel, Hash}
+import ch.epfl.pop.pubsub.graph.{ErrorCodes, GraphMessage, PipelineError}
+import ch.epfl.pop.pubsub.{AskPatternConstants, ClientActor, PubSubMediator}
+import ch.epfl.pop.storage.DbActor
+import ch.epfl.pop.model.network.MethodType
+import ch.epfl.pop.pubsub.graph.validators.RpcValidator
+import ch.epfl.pop.model.network.method.message.Message
+
+import scala.collection.mutable
+import akka.stream.FlowShape
+import ch.epfl.pop.model.network.JsonRpcResponse
+import ch.epfl.pop.model.network.ResultObject
+
+import scala.collection.immutable.HashMap
+import scala.concurrent.{Await, Future}
+
+object ParamsWithMapHandler extends AskPatternConstants {
+
+  def graph(dbActorRef: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow.fromGraph(GraphDSL.create() {
+    implicit builder: GraphDSL.Builder[NotUsed] =>
+      {
+        import GraphDSL.Implicits._
+
+        /* partitioner port numbers */
+        val portPipelineError = 0
+        val portHeartBeatHandler = 1
+        val portGetMsgsByIdHandler = 2
+        val totalPorts = 3
+
+        /* building blocks */
+        val handlerPartitioner = builder.add(Partition[GraphMessage](
+          totalPorts,
+          {
+            case Right(jsonRpcMessage: JsonRpcRequest) => jsonRpcMessage.getParams match {
+                case _: Heartbeat       => portHeartBeatHandler
+                case _: GetMessagesById => portGetMsgsByIdHandler
+
+              }
+            case _ => portPipelineError // Pipeline error goes directly in handlerMerger
+          }
+        ))
+
+        val heartbeatHandler = builder.add(ParamsWithMapHandler.heartbeatHandler(dbActorRef))
+        val getMessagesByIdHandler = builder.add(ParamsWithMapHandler.getMessagesByIdHandler(dbActorRef))
+
+        val handlerMerger = builder.add(Merge[GraphMessage](totalPorts))
+
+        /* glue the components together */
+        handlerPartitioner.out(portPipelineError) ~> handlerMerger
+        handlerPartitioner.out(portHeartBeatHandler) ~> heartbeatHandler ~> handlerMerger
+        handlerPartitioner.out(portGetMsgsByIdHandler) ~> getMessagesByIdHandler ~> handlerMerger
+
+        /* close the shape */
+        FlowShape(handlerPartitioner.in, handlerMerger.out)
+
+      }
+  })
+
+  def heartbeatHandler(dbActorRef: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map({
+    case Right(jsonRpcMessage: JsonRpcRequest) =>
+      val receivedHeartBeat: Map[Channel, Set[Hash]] = jsonRpcMessage.getParams.asInstanceOf[Heartbeat].channelsToMessageIds // pas sûr
+      val ask = dbActorRef ? DbActor.GetSetOfChannels()
+      val answer = Await.result(ask, duration)
+      val setOfChannels: Set[Channel] = answer.asInstanceOf[DbActor.DbActorGetSetOfChannelsAck].channels.map(s => Channel(s))
+      var localHeartBeat: mutable.HashMap[Channel, List[Hash]] = mutable.HashMap()
```

Comment:
I suggest that you either use `val + mutable` or `var + immutable`,  I personally prefer to stick with the latter as you may have to do conversions from mutable to immutable Map when creating objects or returning them.